### PR TITLE
Rename "modulus" to "remainder"

### PR DESCRIPTION
**Pull request by aaronfranke**
I've renamed "modulus" to "remainder", added clarification for these terms, added clarification for division in C#, ~~added a tip for how to use `&`~~, and made a few other minor tweaks.

Part of fixing https://github.com/dotnet/docs/issues/4827 but I'm not sure what to do about the Decimal docs: https://github.com/dotnet/docs/issues/4827#issuecomment-378013490

### CODE REVIEW

File: .openpublishing.redirection.json [hunk 1]
```diff
@@ -329,6 +329,14 @@
             "source_path": "docs/csharp/interop.md",
             "redirect_url": "/dotnet/csharp/programming-guide/interop/index"
         },
+        {
+            "source_path": "docs/csharp/language-reference/operators/modulus-assignment-operator.md",
+            "redirect_url": "/dotnet/csharp/language-reference/operators/remainder-assignment-operator.md"
```
Comment (pkulikov) [inline .openpublishing.redirection.json]: `redirect_url` doesn't need `.md` suffix

File: .openpublishing.redirection.json [hunk 2]
```diff
@@ -329,6 +329,14 @@
...
+        },
+        {
+            "source_path": "docs/csharp/language-reference/operators/modulus-operator.md",
+            "redirect_url": "/dotnet/csharp/language-reference/operators/remainder-operator.md"
```
Comment (pkulikov) [inline .openpublishing.redirection.json]: Same as above: remove `.md` suffix from the `redirect_url`

File: docs/csharp/language-reference/operators/and-operator.md [hunk 3]
```diff
@@ -18,19 +18,21 @@ author: "BillWagner"
...
   
- The `&` operator evaluates both operators regardless of the first one's value. For example:  
+ The binary `&` operator evaluates both operators regardless of the first one's value, in contrast to the [conditional AND operator](../../../csharp/language-reference/operators/conditional-and-operator.md) `&&`. For example:  
   
  [!code-csharp[csRefOperators#37](../../../csharp/language-reference/operators/codesnippet/CSharp/and-operator_1.cs)]  
   
  User-defined types can overload the binary `&` operator (see [operator](../../../csharp/language-reference/keywords/operator.md)). Operations on integral types are generally allowed on enumeration. When a binary operator is overloaded, the corresponding assignment operator, if any, is also implicitly overloaded.  
   
+ The `&` operator can also be used as extremely efficient modulus for power-of-two divisors. `45 & 15` is equivalent to `45 % 16`, but as `%` is the [remainder operator](../../../csharp/language-reference/operators/remainder-operator.md), the results are different for negative dividends.
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/and-operator.md]: I don't think this paragraph is necessary here. Performance-focused folks know this fact, for others it might be more confusing. In general it's not description of *what* the operator does (the purpose of this article), but *what for* it might be used.

If you still want to keep this paragraph, please make it more clear:
- what is *extremely efficient*? Doesn't the compiler already optimize `a % b` to use bit mask when b is the power-of-two? What about [this comment](https://github.com/dotnet/csharplang/issues/1406#issuecomment-375085534)? If that's true, no need to use `&` for `%` at all.
- what definition of modulus do you use?
- looks like a typo: *`45 & 15` is equivalent to `45 % 16`*. Of course, it's not a typo but bit tricks. That also should be explained, as bit tricks might be not common knowledge, especially for the folks who learn about `&` operator.
- what is difference at negative cases?

File: docs/csharp/language-reference/operators/division-operator.md [hunk 4]
```diff
@@ -16,12 +16,14 @@ author: "BillWagner"
...
   
 ## Remarks  
  User-defined types can overload the `/` operator (see [operator](../../../csharp/language-reference/keywords/operator.md)). An overload of the `/` operator implicitly overloads the [/= operator](division-assignment-operator.md).  
   
- When you divide two integers, the result is always an integer. For example, the result of 7 / 3 is 2. To determine the remainder of 7 / 3, use the remainder operator ([%](../../../csharp/language-reference/operators/modulus-operator.md)). To obtain a quotient as a rational number or fraction, give the dividend or divisor type `float` or type `double`. You can assign the type implicitly if you express the dividend or divisor as a decimal by putting a digit to the right side of the decimal point, as the following example shows.  
+ When you divide two integers, the result is always an integer. For example, the result of 7 / 3 is 2. To determine the remainde, use the [remainder operator](../../../csharp/language-reference/operators/remainder-operator.md) `%`. For integer types, the `/` operator rounds towards zero. This is not to be confused with floored division, which C# does not have.
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/division-operator.md]: Typo in the second sentence: *remainde*
I would put the pre-last sentence to be the second one, and add the corresponding example with `-7 / 3`. Now, the remainder remark is in the middle of the division story.
The last sentence can be removed at all: it doesn't add new information and might be confusing. It's like *Snow is white* vs *Snow is white, not black*

File: docs/csharp/language-reference/operators/division-operator.md [hunk 5]
```diff
@@ -16,12 +16,14 @@ author: "BillWagner"
...
- When you divide two integers, the result is always an integer. For example, the result of 7 / 3 is 2. To determine the remainder of 7 / 3, use the remainder operator ([%](../../../csharp/language-reference/operators/modulus-operator.md)). To obtain a quotient as a rational number or fraction, give the dividend or divisor type `float` or type `double`. You can assign the type implicitly if you express the dividend or divisor as a decimal by putting a digit to the right side of the decimal point, as the following example shows.  
+ When you divide two integers, the result is always an integer. For example, the result of 7 / 3 is 2. To determine the remainde, use the [remainder operator](../../../csharp/language-reference/operators/remainder-operator.md) `%`. For integer types, the `/` operator rounds towards zero. This is not to be confused with floored division, which C# does not have. 
+ 
+ To obtain a quotient as a rational number or fraction, give the dividend or divisor type `float` or type `double`. You can assign the type implicitly if you express the dividend or divisor as a decimal by putting a digit to the right side of the decimal point, as the following example shows.
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/division-operator.md]: I'm not sure *give type* or *assign the type* are widely used phrasings. The word *cast* should be used. 
*...if you express the dividend or divisor as a decimal...* - that might be confused with `System.Decimal`. Floating-point number?

File: docs/csharp/language-reference/operators/multiplication-operator.md [hunk 6]
```diff
@@ -16,11 +16,9 @@ author: "BillWagner"
 ms.author: "wiwagn"
 ---
 # * Operator (C# Reference)
-The multiplication operator (`*`), which computes the product of its operands.  Also, the dereference operator, which allows reading and writing to a pointer.  
+The multiplication operator (`*`) computes the multiplicative product of its operands. All numeric types have predefined multiplication operators.
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/multiplication-operator.md]: *multiplicative* can be removed, as product is about multiplication.
Why did you remove the sentence about dereferencing?

File: docs/csharp/language-reference/operators/remainder-operator.md [hunk 7]
```diff
@@ -8,21 +8,25 @@ ms.topic: "article"
...
 caps.latest.revision: 15
 author: "BillWagner"
 ms.author: "wiwagn"
 ---
 # % Operator (C# Reference)
-The `%` operator computes the remainder after dividing its first operand by its second. All numeric types have predefined remainder operators.  
+The remainder operator (`%`) computes the remainder after dividing its first operand by its second. All numeric types have predefined remainder operators. 
   
 ## Remarks  
+ The formula `a % b` will always return a value on the range `(-b, b)`, exclusive (it can never return `b` or `-b`).
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/remainder-operator.md]: *The formula `a % b` will always return* -> *The `a % b` returns*
Should be more clarification here: there are two possible answers in the range `(-b, b)`. Which one is returned?

File: docs/csharp/language-reference/operators/remainder-operator.md [hunk 8]
```diff
@@ -8,21 +8,25 @@ ms.topic: "article"
...
+  
+ This is not to be confused with canonical modulus, which satisfies a similar rule but with floored division and returns values on the range `[0, b)`. C# does not have an operator for canonical modulus. However, the behavior is the same for positive dividends.
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/remainder-operator.md]: Not sure, if this paragraph is necessary. Let's make the definition above is clear enough not to be confusing.

File: docs/csharp/language-reference/operators/and-operator.md (hunk 3 above)
Comment (aaronfranke) [reply docs/csharp/language-reference/operators/and-operator.md]: I originally had it more explained but I decided it was too long. The compiler does auto-optimize in this way when passed a `uint` but it is uncommon to see `uint` compared to `int`. The negative difference is explained on the `%` page, but perhaps this is not clear.

If you believe this part should be removed, I will.

File: docs/csharp/language-reference/operators/division-operator.md (hunk 5 above)
Comment (aaronfranke) [reply docs/csharp/language-reference/operators/division-operator.md]: I didn't even notice that, a strange sentence indeed. That was there before my commit. Still, I can correct it.

File: docs/csharp/language-reference/operators/multiplication-operator.md (hunk 6 above)
Comment (aaronfranke) [reply docs/csharp/language-reference/operators/multiplication-operator.md]: At first I thought "hey, this probably belongs in the Remarks section", but there is already a (much better written) sentence in the Remarks section for its usage with pointers.

File: docs/csharp/language-reference/operators/remainder-operator.md (hunk 7 above)
Comment (aaronfranke) [reply docs/csharp/language-reference/operators/remainder-operator.md]: No, there are an infinite number of answers. If `b` is `3` then it can return `-2`, `-1`, `0`, `1`, `2`. Or if non-integer, it can also return values like `2.5`.

File: docs/csharp/language-reference/operators/and-operator.md (hunk 3 above)
Comment (pkulikov) [reply docs/csharp/language-reference/operators/and-operator.md]: I think it can be removed.

File: docs/csharp/language-reference/operators/division-operator.md (hunk 5 above)
Comment (pkulikov) [reply docs/csharp/language-reference/operators/division-operator.md]: Oh, indeed, sorry, that sentence was already in the text. Anyway, would be nice, if we can make it better.

File: docs/csharp/language-reference/operators/multiplication-operator.md (hunk 6 above)
Comment (pkulikov) [reply docs/csharp/language-reference/operators/multiplication-operator.md]: Still it should be mentioned in the first paragraph as well, since Remarks is optional to read and one might skip that section.

File: docs/csharp/language-reference/operators/remainder-operator.md (hunk 7 above)
Comment (pkulikov) [reply docs/csharp/language-reference/operators/remainder-operator.md]: I meant two possible outputs for fixed a and b. 5 % 3 returns -1 or 2, both values are in the (-3, 3) range.

Comment (pkulikov) [reply docs/csharp/language-reference/operators/remainder-operator.md]: Though, I admit I've misinterpreted this sentence.

File: docs/csharp/language-reference/operators/and-operator.md (hunk 3 above)
Comment (BillWagner) [reply docs/csharp/language-reference/operators/and-operator.md]: I agree with @pkulikov This paragraph can be removed. It's somewhat off topic.

File: docs/csharp/language-reference/operators/remainder-operator.md (hunk 8 above)
Comment (BillWagner) [reply docs/csharp/language-reference/operators/remainder-operator.md]: I would keep this paragraph in the remarks. It is clarifying for any reader that is trying to make the distinction between remainder and modulus.

File: docs/csharp/language-reference/operators/remainder-operator.md (hunk 7 above)
Comment (BillWagner) [reply docs/csharp/language-reference/operators/remainder-operator.md]: I think the text added is good. I'd change "return a value" to "return an integer value". That ensures that there are only 2 possible values. Then, I'd add:

>  If `q = a / b` and `r = a % b`, then `a = q * b + r`.

That ensures there is exactly one answer that satisfies the definition.

File: docs/csharp/language-reference/operators/and-operator.md [hunk 9]
```diff
@@ -18,19 +18,21 @@ author: "BillWagner"
 ms.author: "wiwagn"
 ---
 # &amp; Operator (C# Reference)
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/and-operator.md]: You've made a great sweep of these articles. Can you update the ms.date metadata value to "04/03/2018"

File: docs/csharp/language-reference/operators/division-operator.md [hunk 10]
```diff
@@ -16,12 +16,14 @@ author: "BillWagner"
 ms.author: "wiwagn"
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/division-operator.md]: Also, update the ms.date value to today.

File: docs/csharp/language-reference/operators/remainder-assignment-operator.md [hunk 11]
```diff
@@ -8,8 +8,8 @@ ms.topic: "article"
 f1_keywords:
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/remainder-assignment-operator.md]: Also, update the ms.date.

File: docs/csharp/language-reference/operators/remainder-operator.md [hunk 12]
```diff
@@ -8,21 +8,25 @@ ms.topic: "article"
 f1_keywords:
```
Comment (BillWagner) [inline docs/csharp/language-reference/operators/remainder-operator.md]: update the ms.date field.

**Review by BillWagner — CHANGES_REQUESTED** [review]
Thanks for making these updates.  We really appreciate it. :clap:

I've reviewed the updates. I agree with the comments made so far, and added a few suggestions.

Just @ - mention us when you've updated it, and I'll review again and merge it.

**Comment by pkulikov** [comment]
@BillWagner @cartermp should those two F# pages be updated with *modulus* to *remainder* changes as well?
https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/symbol-and-operator-reference/
https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/symbol-and-operator-reference/arithmetic-operators

**Comment by aaronfranke** [comment]
@BillWagner @pkulikov I've made several changes, let me know if there are any further concerns.

File: docs/csharp/language-reference/operators/remainder-operator.md [hunk 13]
```diff
@@ -1,28 +1,32 @@
...
 caps.latest.revision: 15
 author: "BillWagner"
 ms.author: "wiwagn"
 ---
 # % Operator (C# Reference)
-The `%` operator computes the remainder after dividing its first operand by its second. All numeric types have predefined remainder operators.  
+The remainder operator (`%`) computes the remainder after dividing its first operand by its second. All numeric types have predefined remainder operators. 
   
 ## Remarks  
+ The formula `a % b` will always return a value on the range `(-b, b)`, exclusive (it can never return `b` or `-b`), keeping the sign of the dividend. The remainder operator satisfies the rule where if `q = a / b` and `r = a % b`, then `a = q * b + r`.
```
Comment (mikedn) [inline docs/csharp/language-reference/operators/remainder-operator.md]: IMO it would be preferable to use the same definition that is used in both the C# spec and the IL spec - `a % b = a - (a / b) * b`.

File: docs/csharp/language-reference/operators/remainder-operator.md [hunk 14]
```diff
@@ -1,28 +1,32 @@
...
+  - "remainder operator [C#]"
   - "% operator [C#]"
 ms.assetid: 3b74f4f9-fd9c-45e7-84fa-c8d71a0dfad7
 caps.latest.revision: 15
 author: "BillWagner"
 ms.author: "wiwagn"
 ---
 # % Operator (C# Reference)
-The `%` operator computes the remainder after dividing its first operand by its second. All numeric types have predefined remainder operators.  
+The remainder operator (`%`) computes the remainder after dividing its first operand by its second. All numeric types have predefined remainder operators.
```
Comment (mikedn) [inline docs/csharp/language-reference/operators/remainder-operator.md]: This is pre-existing text so it's a separate issue but it's worth pointing out that the text "all numeric types have predefined X operators" is kind of bogus. Small numeric types (byte, short etc.) do not have such operators.

File: docs/csharp/language-reference/operators/division-operator.md [hunk 15]
```diff
@@ -16,12 +16,16 @@ author: "BillWagner"
...
   
 ## Remarks  
  User-defined types can overload the `/` operator (see [operator](../../../csharp/language-reference/keywords/operator.md)). An overload of the `/` operator implicitly overloads the [/= operator](division-assignment-operator.md).  
   
- When you divide two integers, the result is always an integer. For example, the result of 7 / 3 is 2. To determine the remainder of 7 / 3, use the remainder operator ([%](../../../csharp/language-reference/operators/modulus-operator.md)). To obtain a quotient as a rational number or fraction, give the dividend or divisor type `float` or type `double`. You can assign the type implicitly if you express the dividend or divisor as a decimal by putting a digit to the right side of the decimal point, as the following example shows.  
+ When you divide two integers, the result is always an integer. For example, the result of 7 / 3 is 2. This is not to be confused with floored division, as the `/` operator rounds towards zero: -7 / 3 is -2.  
+  
+ To obtain a quotient as a rational number, use floating types such as `float` or `double`. This can be done easily by appending `.0` to the end of a number.
```
Comment (pkulikov) [inline docs/csharp/language-reference/operators/division-operator.md]: *floating types* - I'm not sure if that's correct; let's just simplify:
*...use `float` or `double` type. To do that, append `.0` to a number literal as the example below shows.* 

(Append always adds to the end by definition).

**Review by pkulikov — COMMENTED** [review]
I've had just one remark on wording. @aaronfranke could you please also update F# articles mentioned [in the comment above](https://github.com/dotnet/docs/pull/4860#issuecomment-378314754). I guess Visual Basic docs also might contain some *modulus* inclusions.

File: docs/csharp/language-reference/operators/division-operator.md (hunk 15 above)
Comment (mikedn) [reply docs/csharp/language-reference/operators/division-operator.md]: The suggestion to append `.0` is slightly misleading, that will produce a double literal, you need `.0f` to produce a float literal. It may be better to simply state "use floating point types"  instead of going into details about how to produce a floating point literal.

Comment (aaronfranke) [reply docs/csharp/language-reference/operators/division-operator.md]: @mikedn @pkulikov What about "Appending `.0` to a number will make it a `double`." ?

**Comment by aaronfranke** [comment]
So it looks like I should edit the F# docs too, and maybe VB ones? As I'm not familiar with these languages I'd like to confirm they have remainder behavior.

File: docs/csharp/language-reference/operators/division-operator.md (hunk 15 above)
Comment (mikedn) [reply docs/csharp/language-reference/operators/division-operator.md]: I'd say that a documentation page about an operator is not the place to explain how a floating point literal is formed. Simply hinting the user to use floating point should be enough.

But then I do not know the intended target audience of this documentation. Something tells me that it may have been intended for beginners, there are too many cases where some things are over simplified to the point that they're wrong while other things are, not sure what word to use for it, accentuated, redundant etc.

I supposed that for beginners it may be worth reminding them how floating literals are formed, many do not realize that there is a difference between 2, 2.0f and 2.0.

Comment (pkulikov) [reply docs/csharp/language-reference/operators/division-operator.md]: I think in this case we can omit `float` type mention and only refer to `double` as it's about having a quotient non-integer. OR: remove remark on appending: if both operands are not number literals, that remark is not correct and casting should be used.

**Comment by cartermp** [comment]
@aaronfranke Both indeed have remainder behavior.

File: docs/csharp/language-reference/operators/remainder-operator.md (hunk 13 above)
Comment (mikedn) [reply docs/csharp/language-reference/operators/remainder-operator.md]: > BTW, where is that formula in the C# spec?

I'm not sure I understand the question. It seems that you already found it.

File: docs/csharp/language-reference/operators/remainder-operator.md (hunk 14 above)
Comment (BillWagner) [reply docs/csharp/language-reference/operators/remainder-operator.md]: The C# spec says that there is a predefined `%` operator for `int`, `uint`, `long`, `ulong`, `float`, `double`, and `decimal`. 

`byte`, `short`, `ushort`, are all implicitly convertible to `int`, so the `%` operator works and does what's expected:

```csharp
byte b1 = 5;
byte b2 = 2;
int b3 = b1 % b2; // b1 and b2 are both converted to `int`, result is `int`
```

If you were to change it, I'd suggest "All numeric types support the remainder operator". It makes the same point, without disagreeing with the spec.

File: docs/csharp/language-reference/operators/division-operator.md (hunk 15 above)
Comment (BillWagner) [reply docs/csharp/language-reference/operators/division-operator.md]: Good discussion.  I'll suggest not using `.0` at all, as that works on literals, but does not on variables. For the second sentence, I'd say:

> There are many ways to convert between [built in numeric types](../keywords/reference-tables-for-types.md)

Note that I used the internal link in my comment.

**Comment by BillWagner** [comment]
@aaronfranke @cartermp Yes, the VB and F# docs need updating as well. You can add them to this PR, but if you don't feel comfortable in those languages, or don't have time, we'll make an issue and address those changes ourselves.

Thanks for all your work on these docs.

File: docs/csharp/language-reference/operators/division-operator.md (hunk 15 above)
Comment (aaronfranke) [reply docs/csharp/language-reference/operators/division-operator.md]: I think I'll make the first part say ", use the `float`, `double`, or `decimal` types." I was intentionally vague with "floating types" because there is more than just `float` and `double`, as `decimal` exists.

File: docs/csharp/language-reference/operators/remainder-operator.md (hunk 14 above)
Comment (aaronfranke) [reply docs/csharp/language-reference/operators/remainder-operator.md]: @mikedn @BillWagner What about for other operations? Does `b1 + b2` give `byte`? Multiplication? Division? Or is it just Remainder missing an operator definition for `byte`?

**Comment by aaronfranke** [comment]
I have decided to NOT edit the Visual Basic docs as I'm not familiar enough with the language and the situation seems to be a bit more complicated (the operator appears to be the text "Mod" itself).

I was, however, delighted by how easy it was to fix up the F# docs. I also made a few more tweaks as @mikedn @BillWagner @pkulikov suggested. 

I believe this PR is good to merge. (after https://github.com/dotnet/docs/issues/4789 ?)

**Review by pkulikov — APPROVED** [review]
@aaronfranke that looks good to me. Thank you very much for all the effort to spot the issue, discuss it, and fix the docs at the end!

**Review by cartermp — APPROVED** [review]

**Review by BillWagner — APPROVED** [review]
